src/btree/bt_curprev.c

```diff
@@ -385,55 +386,73 @@ __cursor_var_prev(WT_CURSOR_BTREE *cbt, bool newpage, bool restart, size_t *skip
-                if (ins != NULL && WT_INSERT_RECNO(ins) > cbt->recno)
-                    cbt->recno = WT_INSERT_RECNO(ins);
+        if (cbt->cip_saved == cip && F_ISSET(cbt, WT_CBT_CACHEABLE_RLE_CELL)) {
+            cbt->iface.value.data = cbt->tmp->data;
+            cbt->iface.value.size = cbt->tmp->size;
+            return (0);
+        }
 
-                /* Adjust for the outer loop decrement. */
-                ++cbt->recno;
+        /* Unpack the cell and build the return information. */
+        cell = WT_COL_PTR(page, cip);
+        __wt_cell_unpack_kv(session, page->dsk, cell, &unpack);
+        rle = __wt_cell_rle(&unpack);
+        if (unpack.type == WT_CELL_DEL) {
+            if (rle == 1) {
                 ++*skippedp;
                 continue;
             }
 
-            WT_RET(__wt_bt_col_var_cursor_walk_txn_read(session, cbt, page, &unpack, cip));
-            if (cbt->upd_value->type == WT_UPDATE_INVALID ||
-              cbt->upd_value->type == WT_UPDATE_TOMBSTONE) {
-                ++*skippedp;
-                continue;
-            }
-            return (0);
+            /*
+             * There can be huge gaps in the variable-length column-store name space appearing as
+             * deleted records. If more than one deleted record, do the work of finding the next
+             * record to return instead of looping through the records.
+             *
+             * First, find the largest record in the update list that's smaller than the current
+             * record.
+             */
+            ins = __col_insert_search_lt(cbt->ins_head, cbt->recno);
+
+            /*
+             * Second, for records with RLEs greater than 1, the above call to __col_var_search
+             * located this record in the page's list of repeating records, and returned the
+             * starting record. The starting record - 1 is the record to which we could skip, if
+             * there was no larger record in the update list.
+             */
+            cbt->recno = rle_start - 1;
+            if (ins != NULL && WT_INSERT_RECNO(ins) > cbt->recno)
+                cbt->recno = WT_INSERT_RECNO(ins);
+
+            /* Adjust for the outer loop decrement. */
+            ++cbt->recno;
+            ++*skippedp;
+            continue;
+        }
+
+        WT_RET(__wt_txn_read(session, cbt, NULL, cbt->recno, cbt->ins ? cbt->ins->upd : NULL));
+        if (cbt->upd_value->type == WT_UPDATE_INVALID ||
+          cbt->upd_value->type == WT_UPDATE_TOMBSTONE) {
+            ++*skippedp;
+            continue;
         }
-        cbt->iface.value.data = cbt->tmp->data;
-        cbt->iface.value.size = cbt->tmp->size;
+        WT_RET(__wt_value_return(cbt, cbt->upd_value));
+
+        /*
+         * It is only safe to cache the value for other keys in the same RLE cell if it is globally
+         * visible. Otherwise, there might be some older timestamp where the value isn't uniform
```

Comment:
Yes, it can be. What gets written to disk is the latest value and its timestamp, which are uniform across the cell, but the previous values for those records aren't necessarily the same or cover the same timestamp range. This is what the new hs26 exercises. (The problem was originally caught by hs21, but only by luck.)